graphql/query_resolver.go

```diff
@@ -865,101 +864,85 @@ func (r *queryResolver) MainlineCommits(ctx context.Context, options MainlineCom
+	// We will loop through each version returned from GetMainlineCommitVersionsWithOptions and see if there is a commit
+	// that matches the filter parameters (if any).
 	// If there is a match, we will add it to the array of versions to be returned to the user.
-	// If there are not enough matches to satisfy our limit, we will call GetMainlineCommitVersionsWithOptions again with the next order number to check and repeat the process.
+	// If there are not enough matches to satisfy our limit, we will call GetMainlineCommitVersionsWithOptions again
+	// with the next order number to check and repeat the process.
 	for matchingVersionCount < limit {
-		// If we no longer have any more versions to check break out and return what we have.
+		// If we no longer have any more versions to check, break out of the loop.
 		if len(versions) == 0 {
 			break
 		}
-		// If we have checked more versions than the MaxMainlineCommitVersionLimit then break out and return what we have.
+
+		// If we have checked more versions than the MaxMainlineCommitVersionLimit, break out of the loop.
 		if versionsCheckedCount >= model.MaxMainlineCommitVersionLimit {
-			// Return an error if we did not find any versions that match.
 			if matchingVersionCount == 0 {
 				return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("Matching version not found in %d most recent versions", model.MaxMainlineCommitVersionLimit))
 			}
 			break
 		}
-		versionsCheckedCount++
-		v := versions[index]
-		apiVersion := restModel.APIVersion{}
-		apiVersion.BuildFromService(v)
-
-		// If the version was created before we started caching activation status we must manually verify it and cache that value.
-		if v.Activated == nil {
-			err = setVersionActivationStatus(ctx, &v)
-			if err != nil {
-				return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error setting version activation status: %s", err.Error()))
-			}
-		}
-		mainlineCommitVersion := MainlineCommitVersion{}
-		shouldCollapse := false
-		if !utility.FromBoolPtr(v.Activated) {
-			shouldCollapse = true
-		} else if isPopulated(buildVariantOptions) && utility.FromBoolPtr(options.ShouldCollapse) {
+
+		var hasMatchingTasksMap map[string]bool
+		if hasFilters {
+			activeVersions := utility.FilterSlice(versions, func(s model.Version) bool { return utility.FromBoolPtr(s.Activated) })
 			opts := task.HasMatchingTasksOptions{
 				TaskNames:                  buildVariantOptions.Tasks,
 				Variants:                   buildVariantOptions.Variants,
 				Statuses:                   getValidTaskStatusesFilter(buildVariantOptions.Statuses),
 				IncludeNeverActivatedTasks: true,
 			}
-			hasTasks, err := task.HasMatchingTasks(ctx, v.Id, opts)
-			if err != nil {
-				return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error checking if version has tasks: %s", err.Error()))
-			}
-			if !hasTasks {
-				shouldCollapse = true
+			if hasMatchingTasksMap, err = concurrentlyBuildHasMatchingTasksMap(ctx, activeVersions, opts); err != nil {
```

Comment:
nit -- preference to not assign values in if statements that you don't intend to use outside of the if statement; I would break this down
```
hasMatchingTasksMap, err = concurrentlyBuildHasMatchingTasksMap(ctx, activeVersions, opts)
if err != nil {  ....
```